pkg/upgrade/key_pair.go

```diff
@@ -8,42 +8,55 @@ import (
 	"strings"
 
 	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime"
-	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/clientcmd"
 	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/services/certificates"
-	clusterapiv1alpha1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+	clusterv1alpha1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
 )
 
-// KeyPair contains a cert and key.
-type KeyPair struct {
+// KubeconfigSecretKey is the key where the kubeconfig is stored in the secret.
+const KubeconfigSecretKey = "kubeconfig"
+
+// keyPair contains a cert and key.
+type keyPair struct {
 	Cert []byte `json:"cert"`
 	Key  []byte `json:"key"`
 }
 
-type keyPairGetter struct {
-	secretClient corev1client.SecretsGetter
+// secrets has already been scoped to namespace, so, .Secrets("my-namespace") returns one of these.
+type secrets interface {
+	Get(string, metav1.GetOptions) (*v1.Secret, error)
 }
 
-func newKeyPairGetter(secretClient corev1client.SecretsGetter) *keyPairGetter {
-	return &keyPairGetter{
-		secretClient: secretClient,
+// NewRestConfigFromKubeconfigSecretRef decodes the kubeconfig stored in a secret and builds a *rest.Config with it.
+func NewRestConfigFromKubeconfigSecretRef(secrets secrets, name string) (*rest.Config, error) {
+	secret, err := secrets.Get(name, metav1.GetOptions{})
+	if err != nil {
+		return nil, errors.WithStack(err)
 	}
-}
 
-func (k *keyPairGetter) getKeyPair(cluster *clusterapiv1alpha1.Cluster, config KeyPairConfig) (*KeyPair, error) {
-	if config.ClusterField != "" {
-		return getEmbeddedKeyPair(cluster, config.ClusterField)
+	kc, ok := secret.Data[KubeconfigSecretKey]
+	if !ok {
+		return nil, errors.Errorf("item 'kubeconfig' not found in secret %q", name)
 	}
-
-	return k.getSecretRefKeyPair(cluster.Namespace, config.SecretRef)
+	out := make([]byte, base64.StdEncoding.DecodedLen(len(kc)))
+	n, err := base64.StdEncoding.Decode(out, kc)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	out = out[:n]
+	return clientcmd.RESTConfigFromKubeConfig(out)
 }
 
-func getEmbeddedKeyPair(cluster *clusterapiv1alpha1.Cluster, path string) (*KeyPair, error) {
-	pathParts := strings.Split(path, ".")
+// NewRestConfigFromClusterField returns a rest.Config configured with the CA key pair found in the cluster's
+// object in the fieldpath specified. For example, "spec.providerSpec.value.caKeyPair" traverses the cluster
+// object going through each '.' delimited field.
+func NewRestConfigFromClusterField(cluster *clusterv1alpha1.Cluster, fieldPath, apiEndpoint string) (*rest.Config, error) {
```

Comment:
Should we call this `NewRestConfigFromCAClusterField`?

Reply:
sure, seems a little more clear